Sources/SwiftFusion/Inference/ValuesStorage.swift

```diff
@@ -0,0 +1,210 @@
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  final func add(_ otherStart: UnsafeRawPointer) {
+    vectorImplementation.add_(otherStart)
+  }
+  
+  /// Scales each element of `self` by `scalar`.
+  final func scale(by scalar: Double) {
+    vectorImplementation.scale_(by: scalar)
+  }
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  final func dot(_ other: UnsafeRawPointer) -> Double {
+    vectorImplementation.dot_(other)
+  }
+}
+
+/// Contiguous storage of homogeneous `EuclideanVector` values of statically unknown type.
+protocol AnyVectorStorageImplementation:
+  AnyVectorStorage, AnyDifferentiableStorageImplementation
+{
+  /// Adds the vector starting at `otherStart` to `self`.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  func add_(_ otherStart: UnsafeRawPointer)
+  
+  /// Scales each element of `self` by `scalar`.
+  func scale_(by scalar: Double)
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  func dot_(_ otherStart: UnsafeRawPointer) -> Double
+}
+
+/// APIs that depend on `EuclideanVector` `Element` type.
+extension ArrayStorageImplementation where Element: EuclideanVector {
+  /// Adds the vector starting at `otherStart` to `self`.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count as `self`.
+  func add(_ otherStart: UnsafePointer<Element>) {
+    withUnsafeMutableBufferPointer { elementBuffer in
+      var otherPointer = otherStart
+      for elementIndex in elementBuffer.indices {
+        elementBuffer[elementIndex] += otherPointer.pointee
+        otherPointer = otherPointer.advanced(by: 1)
+      }
+    }
+  }
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count as `self`.
+  func dot(_ otherStart: UnsafePointer<Element>) -> Double {
+    var result = Double(0)
+    withUnsafeMutableBufferPointer { elementBuffer in
+      var otherPointer = otherStart
+      for elementIndex in elementBuffer.indices {
+        result += elementBuffer[elementIndex].dot(otherPointer.pointee)
+        otherPointer = otherPointer.advanced(by: 1)
+      }
+    }
+    return result
```

Comment:
I think long variable names are actually harming comprehensibility here.  It's just hard to take in what's happening.
```suggestion
    withUnsafeMutableBufferPointer { b in
      var p = otherStart
      for i in b.indices {
        result += b[i].dot(otherPointer.pointee)
        p = p.advanced(by: 1)
      }
    }
    return result
```

Then

```suggestion
    withUnsafeMutableBufferPointer { b in
      var p = otherStart
      for i in b.indices {
        result += b[i].dot(otherPointer.pointee)
        p = p + 1
      }
    }
    return result
```

Then

```suggestion
    withUnsafeMutableBufferPointer { b in
      for i in b.indices {
        result += b[i].dot(otherStart[i])
      }
    }
    return result
```

But see below.